src/core.jl

```diff
@@ -135,3 +138,68 @@ stagedfunction Base.getindex{T,N,D,names,Ax,AxElt}(A::AxisArray{T,N,D,names,Ax,A
 
     return :(A[$(idxs...)])
 end
+
+### Indexing with the element type of the axes ###
+
+abstract AxisType
+immutable Dimensional <: AxisType end
+immutable Categorical <: AxisType end
+immutable Unsupported <: AxisType end
+
+axistype(v::Any) = error("axes must be vectors of concrete types; $(typeof(v)) is not supported")
+axistype(v::AbstractVector) = axistype(eltype(v))
+axistype(T::Type) = Unsupported
+axistype(T::Type{Int}) = Unsupported # Ints are exclusively for real indexing
+axistype{T<:Union(Number, Dates.AbstractTime)}(::Type{T}) = Dimensional
+axistype{T<:Union(Symbol, AbstractString)}(::Type{T}) = Categorical
+
+checkaxis(ax) = checkaxes(axistype(ax), ax)
+checkaxis(::Type{Unsupported}, ax) = nothing # TODO: warn or error?
+# Dimensional axes must be monotonically increasing
+checkaxis{T}(::Type{Dimensional}, ax::Range{T}) = step(ax) > zero(T) || error("Dimensional axes must be monotonically increasing")
+checkaxis(::Type{Dimensional}, ax) = issorted(ax, lt=(<=)) || error("Dimensional axes must be monotonically increasing")
+# Categorical axes must simply be unique
+function checkaxis(::Type{Categorical}, ax)
+    seen = Set(eltype(ax))
+    for elt in ax
+        elt in seen && error("Categorical axes must be unique")
+        push!(seen, elt)
+    end
+end
+
+# A very primitive interval type
+type Interval{T}
+    lo::T
+    hi::T
+    Interval(lo, hi) = lo <= hi ? new(lo, hi) : error("lo must be less than or equal to hi")
+end
+Interval{T}(a::T,b::T) = Interval{T}(a,b)
+Base.promote_rule{T,S}(::Type{Interval{T}}, ::Type{Interval{S}}) = Interval{promote_type(T,S)}
+Base.promote_rule{T,S}(::Type{Interval{T}}, ::Type{S}) = Interval{promote_type(T,S)}
+Base.convert{T,S}(::Type{Interval{T}}, x::Interval{S}) = (R = promote_type(T,S); Interval{R}(convert(R,x.lo),(convert(R,x.hi))))
+Base.convert{T}(::Type{Interval{T}}, x) = Interval(x,x)
+Base.isless(a::Interval, b::Interval) = isless(a.hi, b.lo)
+Base.isless(a::Interval, b) = isless(promote(a,b)...)
+Base.isless(a, b::Interval) = isless(promote(a,b)...)
+
+# Default axes indexing throws an error
+axisindexes(ax, idx) = axisindexes(axistype(ax), ax, idx)
+axisindexes(::Type{Unsupported}, ax, idx) = error("elementwise indexing is not supported for axes of type $(typeof(ax))")
+# Dimensional axes may be indexed by intervals of their elements
+axisindexes{T}(::Type{Dimensional}, ax::AbstractVector{T}, idx::Interval{T}) = searchsorted(ax, idx)
+# Categorical axes may be indexed by their elements
+function axisindexes{T}(::Type{Categorical}, ax::AbstractVector{T}, idx::T)
+    i = findfirst(ax, idx)
+    i == 0 && error("index $idx not found")
+    i
+end
```

Comment:
Seems reasonable.  You can go ahead and add it to this branch.  I'll work tonight on fixing the getindex ambiguities.  It's a bit of a mess.